File: EnvMonitorAndroid/app/src/main/java/com/envmonitor/api/ApiService.kt
package com.envmonitor.api

import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

/**
 * Interface API Service cho giao tiếp với máy chủ
 */
interface ApiService {
    
    /**
     * Lấy dữ liệu hiện tại từ tất cả các cảm biến
     * 
     * @return Map chứa dữ liệu hiện tại
     */
    @GET("/api/current")
    suspend fun getCurrentData(): Map<String, Any>
    
    /**
     * Lấy dữ liệu lịch sử từ tất cả các cảm biến
     * 
     * @param hours Số giờ cần lấy dữ liệu lịch sử
     * @return Map chứa dữ liệu lịch sử
     */
    @GET("/api/historical")
    suspend fun getHistoricalData(@Query("hours") hours: Int = 1): Map<String, List<Map<String, Any>>>
    
    /**
     * Lấy dữ liệu lịch sử cho một thông số cụ thể
     * 
     * @param paramName Tên của thông số cần lấy dữ liệu
     * @param hours Số giờ cần lấy dữ liệu lịch sử
     * @return Danh sách các điểm dữ liệu lịch sử
     */
    @GET("/api/historical/{param}")
    suspend fun getParameterHistoricalData(
        @Path("param") paramName: String,
        @Query("hours") hours: Int = 1
    ): List<Map<String, Any>>
    
    /**
     * Kiểm tra trạng thái kết nối với ThingsBoard
     * 
     * @return Map chứa thông tin trạng thái kết nối
     */
    @GET("/api/status")
    suspend fun getThingsboardStatus(): Map<String, Boolean>
}

File: env_monitor_app/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import '../models/current_data.dart';
import '../models/historical_data.dart';

class ApiService {
  // API base URL - sử dụng URL của Replit hoặc server của bạn
  // Lưu ý: URL này cần được thay đổi khi triển khai
  static const String baseUrl = 'https://a7dc8ca9-0605-486f-b306-92aeef424639-00-2jq9zk4avsodf.janeway.replit.dev:5000';
  
  // Singleton instance
  static final ApiService _instance = ApiService._internal();
  
  factory ApiService() {
    return _instance;
  }
  
  ApiService._internal();

  /// Lấy dữ liệu hiện tại từ server
  Future<CurrentData> getCurrentData() async {
    final response = await http.get(Uri.parse('$baseUrl/api/current'));
    
    if (response.statusCode == 200) {
      return CurrentData.fromJson(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load current data');
    }
  }

  /// Lấy dữ liệu lịch sử từ server
  Future<HistoricalData> getHistoricalData({int hours = 1}) async {
    final response = await http.get(
      Uri.parse('$baseUrl/api/historical${hours > 1 ? "?hours=$hours" : ""}')
    );
    
    if (response.statusCode == 200) {
      return parseHistoricalData(jsonDecode(response.body));
    } else {
      throw Exception('Failed to load historical data');
    }
  }

  /// Lấy dữ liệu lịch sử cho một thông số cụ thể
  Future<List<HistoricalDataPoint>> getParameterHistoricalData(
    String paramName, {
    int hours = 1,
  }) async {
    final response = await http.get(
      Uri.parse('$baseUrl/api/historical/$paramName${hours > 1 ? "?hours=$hours" : ""}')
    );
    
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body) as List;
      return data
          .map((item) => HistoricalDataPoint.fromJson(item as Map<String, dynamic>))
          .toList();
    } else {
      throw Exception('Failed to load parameter historical data');
    }
  }

  /// Kiểm tra trạng thái kết nối với ThingsBoard
  Future<bool> checkThingsboardStatus() async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/api/status'));
      
      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['connected'] as bool;
      }
      return false;
    } catch (e) {
      return false;
    }
  }
}